dali/ft/filecopy.cpp

```diff
@@ -2658,265 +2638,34 @@ void FileSprayer::pushParts()
-        CDateTime dateTime;
-        status = getProgress(dateTime, outputLength);
-        switch (status)
-        {
-            case ApiCopyStatus::Pending:
-                try
-                {
-                    if (blobCopyOp.HasValue() && blobCopyOp.Value().CopyId.HasValue())
-                        blobClient->AbortCopyFromUri(blobCopyOp.Value().CopyId.Value().c_str());
-                    else // not sure that this can ever happen
-                        IERRLOG("AzureBlobClient::AbortCopy() failed: CopyId is empty");
-                    status = ApiCopyStatus::Aborted;
-                }
-                catch(const Azure::Core::RequestFailedException& e)
-                {
-                    IERRLOG("Abort copy operation failed: %s (code %d)", e.ReasonPhrase.c_str(), static_cast<int>(e.StatusCode));
-                    status = ApiCopyStatus::Failed;
-                }
-                // fallthrough to delete any file remnants
-            case ApiCopyStatus::Success:
-                // already copied-> need to delete
-                try
-                {
-                    blobClient->Delete();
-                    status = ApiCopyStatus::Aborted;
-                }
-                catch(const Azure::Core::RequestFailedException& e)
-                {
-                    IERRLOG("Delete operation failed: %s (code %d)", e.ReasonPhrase.c_str(), static_cast<int>(e.StatusCode));
-                    status = ApiCopyStatus::Failed;
-                }
-                break;
-        }
-        return status;
-    }
-};
-#endif
-
 void FileSprayer::transferUsingAPI()
 {
-#ifdef _USE_AZURE
-    // N.B. Only call this member function, if FileSprayer::useAPICopy() has returned true
     // Source
-    StringBuffer sourceClusterName, sourceBaseURI, sourceSAStoken;
+    StringBuffer sourceClusterName;
     distributedSource->getClusterName(0, sourceClusterName);
     Owned<IStoragePlane> sourcePlane = getDataStoragePlane(sourceClusterName.str(), false);
-    CApiCopyClient::buildStorageURIBase(sourcePlane, sourceBaseURI);
-    sourcePlane->getSASToken(sourceSAStoken);
+    Owned<IStorageApiInfo> sourceAPIInfo = sourcePlane->getStorageApiInfo();
 
     // Target
-    StringBuffer targetClusterName, targetBaseURI, targetSAStoken;
+    StringBuffer targetClusterName;
     distributedTarget->getClusterName(0, targetClusterName);
     Owned<IStoragePlane> targetPlane = getDataStoragePlane(targetClusterName.str(), false);
-    CApiCopyClient::buildStorageURIBase(targetPlane, targetBaseURI);
-    targetPlane->getSASToken(targetSAStoken);
-    StorageType targetType = targetPlane->getStorageType();
+    Owned<IStorageApiInfo> targetAPIInfo = targetPlane->getStorageApiInfo();
+
+    // useAPICopy() should have ensured this cannot happen
+    dbgassertex(sourceAPIInfo!=nullptr && targetAPIInfo!=nullptr);
+
+    Owned<IAPICopyClient> copyClient = createApiCopyClient(sourceAPIInfo, targetAPIInfo);
+    if (!copyClient)
```

Comment:
Debatable.  I'm not sure this is right - the api for source and target might be present but incompatible.  E.g. azure and aws.  (I suspect it should also use UERRLOG rather than OERRLOG, especially if operator errors start being reported to eclwatch - It is worth generating some kind of warning though.)
I think if it used UWARNLOG then the change would be fine as it stands.

(Another possibility would be to return IAPICopyClient * from useAPICopy() and pass that into this function.)